# GOTHiC: Gene Ontology Topology from Hi-C

GOTHiC is a python tool that allows for the creation, annotation, and analysis of Hi-C networks. In essence, it leverages graph theory techniques to allow for the assessment of structure-function relationships in nuclear chromatin.

![A summary of GOTHiC's primary workflow](https://github.com/LavalleeAdamLab/GOTHiC/assets/61287366/7d13adaa-5a70-4c1b-b0a7-7116a9736b1b)

## Using GOTHiC:
To use GOTHiC, copy `gothic.py` into a directory within your python PATH, and simply 'from gothic import *'. Individual functions are available for specific tasks involving the conversion of .sam or .hic files to Hi-C graphs, graph functional annotation with Gene Ontology terms or REACTOME metabolic pathways, clustering analysis, and visualization. Function documentation can be found in the Docs/ directory, and scripts are available in the Scripts/ directory for running the most typical workflow. These scripts can be used for Hi-C network creation and normalization from SAM file, graph annotation with Gene Ontology Terms, clustering analysis, A/B compartment analysis, and visualization.  

Dependencies:

graph-tool (install [here](https://graph-tool.skewed.de/installation.html))

required packages: pandas numpy matplotlib scipy tqdm plotly goatools seaborn fanc iced gprofiler sklearn markov_clustering pycairo

required files: download from https://drive.google.com/drive/folders/1ftdomTXNynh5iZewhiUpyHiWQeuT5I9p?usp=sharing ,
or use your preferred versions of GO-obo, gencode, and uniprot IDmapping files. Just be sure to explicitly name these files in your run scripts accordingly.  

![Heatmap visualization of A/B compartments and Network visualization of GO terms with GOTHiC](https://github.com/user-attachments/assets/45a5e6b3-aba6-4c50-97c8-3112fcdb055e)

The recommended entry point for using GOTHiC is to start with either a pre-aligned pair of Hi-C .sam files or a .hic file produced by either FAN-C or Juicer. There are two available binning methods for graph creation: fixed-width binning and custom region binning. For fixed-width binning, simply specify the size each bin should be in base pairs using the binsize argument. If you wish to use custom regions as bins (e.g. TAD boundaries), instead provide a tab delimited regions file where each row is a bin/region and has three columns for chromosome (e.g. chr1, chr2, etc.), within-chromosome start position, and within-chromosome end position. 

For a typical analysis of GO term clustering in a Hi-C dataset as presented in Nygard et al. 2025, run the following scripts in the Scripts/ folder in this order using your input files and parameters of choice:
- fullRunTPD.py
- fullRunSpectralMCL.py
- tableReformat.py
- generateFigures.py

To do the network visualization using Hi-C networks generated by GOTHiC, import the .graphml file into Cytoscape, and select nodes using a file generated by get_nodes_with_term() for your network and desired GO term.




